coderd/notifications/dispatch/webhook_test.go

```diff
@@ -0,0 +1,144 @@
+package dispatch_test
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/http/httptest"
+	"net/url"
+	"testing"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/stretchr/testify/require"
+
+	"cdr.dev/slog"
+	"cdr.dev/slog/sloggers/slogtest"
+	"github.com/coder/serpent"
+
+	"github.com/coder/coder/v2/coderd/notifications/dispatch"
+	"github.com/coder/coder/v2/coderd/notifications/types"
+	"github.com/coder/coder/v2/codersdk"
+	"github.com/coder/coder/v2/testutil"
+)
+
+func TestWebhook(t *testing.T) {
+	t.Parallel()
+
+	const (
+		titleTemplate = "this is the title ({{.Labels.foo}})"
+		bodyTemplate  = "this is the body ({{.Labels.baz}})"
+	)
+
+	msgPayload := types.MessagePayload{
+		Version:          "1.0",
+		NotificationName: "test",
+		Labels: map[string]string{
+			"foo": "bar",
+			"baz": "quux",
+		},
+	}
+
+	tests := []struct {
+		name          string
+		serverURL     string
+		serverTimeout time.Duration
+		serverFn      func(uuid.UUID, http.ResponseWriter, *http.Request)
+
+		expectSuccess   bool
+		expectRetryable bool
+		expectErr       string
+	}{
+		{
+			name: "successful",
+			serverFn: func(msgID uuid.UUID, w http.ResponseWriter, r *http.Request) {
+				var payload dispatch.WebhookPayload
+				err := json.NewDecoder(r.Body).Decode(&payload)
+				require.NoError(t, err)
+				require.Equal(t, "application/json", r.Header.Get("Content-Type"))
+				require.Equal(t, msgID, payload.MsgID)
+				require.Equal(t, msgID.String(), r.Header.Get("X-Message-Id"))
+
+				w.WriteHeader(http.StatusOK)
+				_, err = w.Write([]byte(fmt.Sprintf("received %s", payload.MsgID)))
+				require.NoError(t, err)
+			},
+			expectSuccess: true,
+		},
+		{
+			name: "invalid endpoint",
+			// Build a deliberately invalid URL to fail validation.
+			serverURL:       "invalid .com",
+			expectSuccess:   false,
+			expectErr:       "invalid URL escape",
+			expectRetryable: false,
+		},
+		{
+			name:            "timeout",
+			serverTimeout:   time.Millisecond,
+			expectSuccess:   false,
+			expectRetryable: true,
+			expectErr:       "request timeout",
+		},
+		{
+			name: "non-200 response",
+			serverFn: func(_ uuid.UUID, w http.ResponseWriter, r *http.Request) {
+				w.WriteHeader(http.StatusInternalServerError)
+			},
+			expectSuccess:   false,
+			expectRetryable: true,
+			expectErr:       "non-2xx response (500)",
+		},
```

Comment:
Perhaps we could add a test that sends bad data (wrong format) too, just to have the case covered?

Reply:
It's not actually possible, the webhook dispatch sends a well-formed request.

See `func (w *WebhookHandler) dispatch`